### PR TITLE
ENH: Add group transformers

### LINKED ISSUE #50: Migrate transformers from AFQ-Insight
AFQ-Insight has a few functions and transformer classes to select, remove, manipulate individual groups or subsets of groups in a feature matrix. There is nothing tractometry specific about this so it should go in groupyr.

We can capture all of the functionality in `afqinsight.transform` by adding the following transfomer classes

- [x] `GroupExtractor`: to select and return certain groups from the feature matrix
- [x] `GroupRemover`: to remove certain groups from the feature matrix
- [x] `GroupShuffler`: to shuffle certain groups within a feature matrix without touching the other groups
- ~~`GroupConcatenator`: to concatenate feature matrices with different groups~~

@arokem, @mnarayan: your thoughts?

### PR DESCRIPTION
**Pull request by richford**
Resolves #50 

WIP until I add all of the transformers.

Also, these transformers currently take indices of groups to extract/remove. We should optionally allow the user to pass a feature name to select or remove. We could then check the feature name against elements of the feature column names (provided as another kwarg or taken from `DataFrame.columns` if `X` is a dataframe). That way, the user wouldn't have to remember which group index corresponds to the CST, for example.

### CODE REVIEW

**Comment by richford** [comment]
@arokem, ready to review.

File: groupyr/tests/test_transform.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    gs = GroupShuffler(groups=groups, select=select, random_state=0)
+    X_tr = gs.fit_transform(X)
+    assert_shuffle_match(X, X_tr, groups=groups, select=select, random_state=0)
+
+    gs = GroupShuffler()
+    X_tr = gs.fit_transform(X)
+    assert np.allclose(X_tr, X)  # nosec
+
+
+@pytest.mark.parametrize("Transformer", [GroupExtractor, GroupRemover, GroupShuffler])
```
Comment (arokem) [inline groupyr/tests/test_transform.py:258]: Nice

File: groupyr/transform.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        self.group_names = group_names
+        self.copy_X = copy_X
+
+    def transform(self, X, y=None):
+        """Transform the input data, extracting the desired groups.
+
+        Parameters
+        ----------
+        X : numpy.ndarray
+            The feature matrix
```
Comment (arokem) [inline groupyr/transform.py:123]: What about `y`?

Comment (arokem) [reply groupyr/transform.py:123]: In particular, it looks like `y` is never used. Is it here just for compatibility with the sklearn API?

File: groupyr/transform.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+        )
+        groups = check_groups(groups=self.groups_, X=X, allow_overlap=True)
+        if self.select_ is None:
+            return X
+
+        idx = np.concatenate([groups[e] for e in self.select_])
+        return X[:, idx]
+
+    def fit(self, X=None, y=None):
+        """Learn the groups and number of features from the input data."""
```
Comment (arokem) [inline groupyr/transform.py:139]: Same here. Is `y` an input just for API compatibility?

**Review by arokem — APPROVED** [review]
Looks great. Documentation of these features (e.g., in an example) might be helpful, but can follow on a separate PR.

File: groupyr/transform.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+import numpy as np
+
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils import check_array, check_random_state
+from sklearn.utils import shuffle as util_shuffle
+
+from .utils import check_groups
+
+
+def isiterable(obj):
```
[the remark this answers is not in the document]
Comment (richford) [reply groupyr/transform.py:11]: I know right!

File: groupyr/transform.py (hunk 2 above)
Comment (richford) [reply groupyr/transform.py:123]: Yep. `y=None` is required to be able to use it as part of a pipeline that mixes supervised transformers/estimators that do accept a `y`. See https://scikit-learn.org/stable/developers/develop.html#fitting for more details if interested.

File: groupyr/transform.py (hunk 3 above)
Comment (richford) [reply groupyr/transform.py:139]: Yep. See comment above.
